### PR TITLE
[2.7] bpo-25083: Python 2 can sometimes create incorrect .pyc files

### PR DESCRIPTION
**Pull request by tzickel**
Python 2 never checked for I/O error when reading .py files and
thus could mistake an I/O error for EOF and create incorrect .pyc
files. This adds an check for this and aborts on an error.

https://bugs.python.org/issue25083

### CODE REVIEW

**Comment by tzickel** [comment]
If someone can add / remove a label to update the the-knights-who-say-ni bot.

**Comment by encukou** [comment]
This looks good except for the use of `E_EOF`. That should be reserved for end-of-file, but in this case here we're *not* getting to the end of file.
I don't see a good `E_` constant for read errors in the current list. Could you add a new one in `Include/errcode.h` and `Python/pythonrun.c`, so that this gets a proper error message?

**Comment by encukou** [comment]
And don't worry about the backport title; that's for the person who'll merge this.

**Comment by tzickel** [comment]
Actually not doing this E_EOF causes this to behave badly...

**Comment by tzickel** [comment]
Done.

**Comment by tzickel** [comment]
Here is a script to check for the bug:
https://gist.github.com/tzickel/2ab09100ea21ca0e1e6709c337194bff
needs a recent strance (ubuntu 18.04 or debian sed is fine), and if run inside a docker, the docker needs to be run with --cap-add SYS_PTRACE for strace to work.

File: Parser/tokenizer.c
Comment (vstinner) [inline Parser/tokenizer.c]: Hum, should we call clearerr() here? In case if fp is used again.

**Review by vstinner — CHANGES_REQUESTED** [review]

**Comment by tzickel** [comment]
I have made the requested changes; please review again

**Review by vstinner — APPROVED** [review]
LGTM, but I would prefer that someone else merge it.

**Comment by tzickel** [comment]
@encukou I'm waiting for the convincing part :) I think it's best to merge this ASAP (if it's plan to be used), so it has maximum exposure before a new release.

**Pull request merged**
